### DISCUSSION TITLE
Having problem with Viewer

### DISCUSSION THREAD

**Discussion by RadheChandu** [usnistgov/fipy#943] [Q&A] [answered]
Hello FipyTeam,

I am getting the following error while I was trying the examples on FiPy website. 

```python
Uncaught exception:
Traceback (most recent call last):
  File "~\test1.py", line 30, in <module>
    viewer = Viewer(vars=phi,
  File "C:\ProgramData\Anaconda3\lib\site-packages\fipy\viewers\__init__.py", line 105, in Viewer
    for name, ep in sorted(enpts):
TypeError: '<' not supported between instances of 'EntryPoint' and 'EntryPoint'
```

This was the example code I was working on.

```python
from fipy import Grid1D, ExplicitDiffusionTerm, TransientTerm, Viewer, CellVariable

nx = 50
dx = 1.
mesh = Grid1D(nx=nx, dx=dx)

phi = CellVariable(name="solution variable",
                   mesh=mesh,
                   value=0.)

D = 1.

valueLeft = 1
valueRight = 0

phi.constrain(valueRight, mesh.facesRight)
phi.constrain(valueLeft, mesh.facesLeft)

eqX = TransientTerm() == ExplicitDiffusionTerm(coeff=D)


if __name__ == '__main__':
    viewer = Viewer(vars=phi,
                    datamin=0., datamax=1.)
    viewer.plot()
```

I have even tried installing on a new environment, but still was getting the same error.

Can someone please help me solve this error?

**Comment by guyer** [comment]
If you run `python setup.py test` or `python -c "import fipy; fipy.test()`, it should output a block of versions that are being used like
```
python      3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:41) [Clang 15.0.7 ]
fipy        3.4.4+257.g125bb1fe0.dirty
numpy       1.24.3
pysparse    not installed
scipy       1.10.1
matplotlib  3.7.1
mpi4py      3.1.4
petsc4py    3.18.4
pyamgx      not installed
PyTrilinos  not installed
mayavi      4.8.1
gmsh        4.11.1
solver      petsc
```

Please show what these versions are for you.

**Comment by RadheChandu** [comment]
Hello,

I am getting the following error if I execute that command.

error: Multiple top-level packages discovered in a flat-layout: ['terms', 'meshes', 'solvers', 'viewers', 'matrices', 'steppers', 'variables', 'boundaryConditions'].

To avoid accidental inclusion of unwanted files or directories,
setuptools will not proceed with this build.

If you are trying to create a single distribution with multiple packages
on purpose, you should not rely on automatic discovery.
Instead, consider the following options:

1. set up custom discovery (`find` directive with `include` or `exclude`)
2. use a `src-layout`
3. explicitly set `py_modules` or `packages` with a list of names

To find more information, look for "package discovery" on setuptools docs.

----

 And if create an empty folder and execute the command by going in that folder, I am getting these (attached file).
[New Text Document.txt](https://github.com/usnistgov/fipy/files/12472215/New.Text.Document.txt)

**Reply by guyer** [reply]
OK, thank you. I should have warned you about that "flat-layout" error. Running from within an empty directory is the correct solution to that.

[`pkg_resources`](https://setuptools.pypa.io/en/latest/pkg_resources.html) is deprecated, but it works fine for me with both the versions you are running, as well as newer versions of FiPy dependencies. I'll need to try to set up a Windows environment to test there. I will look into updating FiPy to use [`importlib.metadata`](https://docs.python.org/3/library/importlib.metadata.html), but I'd like to reproduce your error to make sure that's even the problem.

**Comment by guyer** [comment]
I've installed the same packages as you on a Windows machine and I still don't get an error.

Out of curiosity, what does
```bash
python -c "import setuptools; print(setuptools.__version__)"
```
report?

Regardless, I'll work on migrating from `pkg_resources` to `import.metadata`

**Reply by RadheChandu** [reply]
Hello,

My setuptools version is 61.2.0

```bash
(base) D:\fipytest>python -c "import setuptools; print(setuptools.__version__)"
61.2.0
```

Also, I've tried this thing here and noticed that its reporting out matplotlib and mayavi twice as shown below.

```bash
(base) D:\fipytest>python -c "import pkg_resources; print(list(pkg_resources.iter_entry_points(group='fipy.viewers')))"
[EntryPoint.parse('matplotlib = fipy.viewers.matplotlibViewer:MatplotlibViewer'), EntryPoint.parse('mayavi = fipy.viewers.mayaviViewer:MayaviClient'), EntryPoint.parse('matplotlib = fipy.viewers.matplotlibViewer:MatplotlibViewer'), EntryPoint.parse('mayavi = fipy.viewers.mayaviViewer:MayaviClient')]

(base) D:\fipytest>
```

Is this the problem? Because I've tried the same thing on an another PC and noticed it just reports mayavi and matplotlib once.

**Reply by guyer** [reply]
Interesting. The current version of setuptools is 68.2.1, but 61.2.0 is only about a year old.

Yes, I think the multiple entries are exactly what's causing the problem. I had already started to wonder if that wasn't the issue, once I started to try to update the code and understood how sorting was working. I'm close to having something you can test. My new code should tolerate multiple entries (for whatever reason that's even happening).

**Comment by guyer** [comment]
Please try https://github.com/usnistgov/fipy/tree/migrate_pkg_resources2importlib

**Reply by RadheChandu** [reply]
Hello,

Thank you for your quick responses.

I've tried this version and I was getting this error now. 

```python
Traceback (most recent call last):

  File C:\ProgramData\Anaconda3\lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File d:\fipytest\test1.py:30 in <module>
    viewer = Viewer(vars=phi,

  File C:\ProgramData\Anaconda3\lib\site-packages\fipy-migrate_pkg_resources2importlib-py3.9.egg\fipy\viewers\__init__.py:108 in Viewer
    enpts = entry_points(group='fipy.viewers')

TypeError: entry_points() got an unexpected keyword argument 'group'
```

**Reply by guyer** [reply, marked as answer]
Ah. That interface apparently wasn't introduced until Python 3.10. Please update and try again.

**Reply by RadheChandu** [reply]
Thank you so much.

Viewer is working fine now.

**Comment by guyer** [comment]
Thanks for bringing this to our attention. The fix has been merged to `main` and will be in the next release.
